kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -425,29 +428,30 @@ void cacheRealmPermsChanged(String realmRef, Set<String> roles, Set<String> perm
 
         Set<Member> members = azg.getMembers();
         if (members != null && !members.isEmpty()) {
-            for (Member member : members) {
+            for (String perm : permissions) {
+              HashSet<String> permKeysToInvalidate = new HashSet<String>();
+              for (Member member : members) {
                 if (member != null && member.isActive() && member.getUserId() != null) {
 		    boolean canSwap = (member.getRole().isAllowed(SiteService.SITE_ROLE_SWAP));
-                    for (String perm : permissions) {
-                        if (perm != null) {
-                            keysToInvalidate.add(makeCacheKey(member.getUserId(), null, perm, azgRef, false));
+                    if (perm != null) {
```

Comment:
Move check to the outer loop? As it seems like a quick shortcut that means you don't need to iterate again on the members.